File: POO/poo11048216/Aulas/aula18.2/JLDE/Lista.java
class Lista
{
    private Item primeiro;
    private Item ultimo;
    
    public Lista()
    {
        this.setPrimeiro(null);
        this.setUltimo(null);
    }
    
    public void setPrimeiro(Item it)
    {
        this.primeiro = it;
    }
    
    public void setUltimo(Item it)
    {
        this.ultimo = it;
    }
    
    public Item getPrimeiro()
    {
        return this.primeiro;
    }
    
    public Item getUltimo()
    {
        return this.ultimo;
    }
    
    public void inserir(Item novo)
    {
        Item ant=null, atu=this.getPrimeiro();
        boolean achei=false;
        
        while(atu!=null && !achei)
        {
            if(atu.getChave()<novo.getChave())
            {
                ant = atu;
                atu = atu.getProximo();
            }
            else
            {
                achei = true;
                if(atu.getChave()==novo.getChave())
                {
                    return;
                }
            }
        }
        if(atu!=null && ant!=null) // inserindo no meio
        {
            ant.setProximo(novo);
            novo.setProximo(atu);
        }
        else if(ant==null) // inserindo no inicio
        {
            if(this.vazia())
            {
                this.setUltimo(novo);
            }
            this.setPrimeiro(novo);
            novo.setProximo(atu);
        }
        else // inserindo no final
        {
            ant.setProximo(novo);
            this.setUltimo(novo);
        }
    }
    
    public Item remover(int ch)
    {
        Item ant=null, atu=this.getPrimeiro();
        if(!this.vazia())
        {
            while(atu!=null && atu.getChave() != ch)
            {
                ant = atu;
                atu = atu.getProximo();
            }
            if(atu!=null && ant!=null) // removendo do meio ou final
            {
                ant.setProximo(atu.getProximo()); // meio ou final
                if(atu == this.getUltimo()) // soh o final
                {
                    this.setUltimo(ant);
                }
            }
            else if(ant==null)// removendo do inicio
            {
                this.setPrimeiro(atu.getProximo());
                if(atu == this.getUltimo())
                {
                    this.setUltimo(null);
                }
            }
        }
        return atu;
    }
    
    public void mostrar()
    {
        Item atu = this.getPrimeiro();
        while(atu!=null)
        {
            System.out.print(atu.getChave() + (atu.getProximo()!=null?" ":"\n"));
            atu = atu.getProximo();
        }
    }
    
    public boolean vazia()
    {
        return this.getPrimeiro()==null && this.getUltimo()==null;
    }
    
}

File: POO/poo11048216/Aulas/aula18.1/LDE/Lista.cpp
#include "Lista.h"

Lista::Lista()
{
    this->setPrimeiro(NULL);
    this->setUltimo(NULL);
}

void Lista::setPrimeiro(Item *it)
{
    this->primeiro = it;
}

void Lista::setUltimo(Item *it)
{
    this->ultimo = it;
}

Item * Lista::getPrimeiro()
{
    return this->primeiro;
}

Item * Lista::getUltimo()
{
    return this->ultimo;
}


void Lista::inserir(Item *novo) throw (ChaveExisteException)
{
    Item *ant=NULL, *atu=this->getPrimeiro();
    bool achei = false;
    
    if(this->vazia())
    {
        this->setPrimeiro(novo);
        this->setUltimo(novo);
        return;
    }
    
    while(atu!=NULL && !achei)
    {
        if(atu->getChave() < novo->getChave())
        {
            ant = atu;
            atu = atu->getProximo();
        }
        else
        {
            achei = true;
        }
    }
    if(atu!=NULL && atu->getChave() == novo->getChave())
    {
        delete novo;
        throw ChaveExisteException();
    }
    else
    {
        if(ant==NULL) // eh o primeiro
        {
            novo->setProximo(atu);
            this->setPrimeiro(novo);
            
        }
        else if(atu==NULL) // eh o ultimo
        {
            ant->setProximo(novo);
            this->setUltimo(novo);
        }
        else // eh o do meio
        {
            ant->setProximo(novo);
            novo->setProximo(atu);
        }
    }
}

Item * Lista::remover(int ch) throw(ChaveInexistenteException, ListaVaziaException)
{
    Item *ant=NULL, *atu = this->getPrimeiro();
    if(!this->vazia())
    {
        while(atu!=NULL && atu->getChave()!=ch)
        {
            ant = atu;
            atu = atu->getProximo();
        }
        if(atu!=NULL && ant!=NULL)
        {
            ant->setProximo(atu->getProximo());
            if(atu == this->getUltimo())
            {
                this->setUltimo(ant);
            }
        }
        else if(ant==NULL)
        {
            if(this->getPrimeiro() == this->getUltimo())
            {
                this->setUltimo(NULL);
            }
            this->setPrimeiro(atu->getProximo());
        }
        else
        {
            throw ChaveInexistenteException();
        }
    }
    else
    {
        throw ListaVaziaException();
    }
    return atu;
}

void Lista::mostrar() throw (ListaVaziaException)
{
    Item *atu=this->getPrimeiro();
    if(this->vazia())
    {
        throw ListaVaziaException();
    }
    while(atu!=NULL)
    {
        std::cout << atu->getChave() << (atu->getProximo()!=NULL?" ":"\n");
        atu = atu->getProximo();
    }
}

bool Lista::vazia()
{
    return this->getPrimeiro() == NULL && this->getUltimo() == NULL;
}